<a href="https://colab.research.google.com/github/fuse999/DS-Unit-1-Sprint-3-Statistical-Tests-and-Experiments/blob/master/module1-statistics-probability-and-inference/Drive_LS_DS_131_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 3 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [1]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2019-09-16 19:44:33--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2019-09-16 19:44:34 (279 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [0]:
# Load and Clean data

In [5]:
import pandas as pd

# Load Data
df = pd.read_csv('house-votes-84.data', 
                 header=None,
                 names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])
print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [7]:
import numpy as np

df = df.replace({'?':np.NaN, 'n':0, 'y':1})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [13]:
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [0]:
dem = df[df['party'] == 'democrat']

rep = df[df['party'] == 'republican']

In [14]:
dem.groupby(['party']).sum()

handicapped-infants  water-project  ...  duty-free  south-africa
party                                         ...                         
democrat                156.0          120.0  ...      160.0         173.0

[1 rows x 16 columns]

In [15]:
rep.groupby(['party']).sum()

handicapped-infants  water-project  ...  duty-free  south-africa
party                                           ...                         
republican                 31.0           75.0  ...       14.0          96.0

[1 rows x 16 columns]

In [0]:
# Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01

In [0]:
from scipy.stats import ttest_1samp

1) Null Hypothesis:

$H_a$: $\bar{x} \neq 0$ - There is non-zero support for the physician-fee-freeze bill among democrats.
 
2) Alternative Hypothesis

$H_0$ support from dem for physician-fee-freeze bill.

3) Confidence Level: 95% or .95

In [22]:
ttest_1samp(rep['duty-free'], 0, nan_policy='omit')

Ttest_1sampResult(statistic=3.9091802389817065, pvalue=0.00013809783981978333)

4) t-statistic: 3.9091802389817065

5) p-value of 0.00013809783981978333

--- 

Conclusion: Due to a p-value of zero I reject the null hypothesis that republican support is Non-zero and conclude that republican support is 0. 

In [0]:
# Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01

1) Null Hypothesis:

$H_a$: $\bar{x} \neq 0$ - There is non-zero support for the physician-fee-freeze bill among democrats.
 
2) Alternative Hypothesis

$H_0$ support from dem for physician-fee-freeze bill.

3) Confidence Level: 95% or .95

In [19]:
ttest_1samp(dem['physician-fee-freeze'], 0, nan_policy='omit')

Ttest_1sampResult(statistic=3.8396428405331067, pvalue=0.0001550891349511097)

4) t-statistic: 3.8396428405331067

5) p-value of 0.0001550891349511097

--- 

Conclusion: Due to a p-value of zero I reject the null hypothesis that republican support is Non-zero and conclude that republican support is 0. 

In [0]:
# Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

1) Null Hypothesis:

$H_0$ -  allmost no difference between religious-groups bill voting from both the Republicans and Democrats.
 
2) Alternative Hypothesis

$H_0$  large difference between religious-groups bill voting from both the Republicans and Democrats.

3) Confidence Level: 95% or .95

In [24]:
from scipy.stats import ttest_ind
ttest_ind(rep['religious-groups'], dem['religious-groups'], nan_policy='omit')

Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)

In [25]:
rep['religious-groups'].mean()

0.8975903614457831

In [26]:
dem['religious-groups'].mean()

0.47674418604651164

4) t-statistic: 9.737575825219457

5) p-value of 0.000000000000000000023936722520597287

--- 

Conclusion: Due to a p-value of zero I confirm the null hypothesis that there is allmost no difference between religious-groups bill voting from both the Republicans and Democrats.. 